### PR TITLE
Wip-enable-nsipsec

### PR DESCRIPTION
**Pull request by JoshSalomon**
Just for test and build (clauster bot) purpose, no need to merge

### CODE REVIEW

**Comment by anuragthehatter** [comment]
@JoshSalomon  As discussed on [slack](https://redhat-internal.slack.com/archives/C06665AQN0J/p1704300034230869), this PR along with [API changes](https://github.com/openshift/api/pull/1667) is disabling ipsec on upgrade.

File: pkg/network/ovn_kubernetes.go [hunk 1]
```diff
@@ -532,10 +535,33 @@ func renderOVNKubernetes(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.Bo
...
+	if conf.IPsecConfig != nil {
+		klog.Infof("IPsec ipsecConfig = %+v", conf.IPsecConfig)
+		if conf.IPsecConfig.Mode != "" {
+			mode = conf.IPsecConfig.Mode
+		} else {
+			mode = operv1.IPsecModeFull // BW compatiniglity with existing configs
+			//TO-REMOVE debug log msg
+			klog.Infof("IPsec mode is not set in ipsecConfig. Assuming upgrade: setting IPsec mode to Full")
+			// For upgrade only - update the object to a valid value
+			conf.IPsecConfig.Mode = operv1.IPsecModeFull
```
Comment (JoelSpeed) [inline pkg/network/ovn_kubernetes.go:557]: No, you shouldn't do this. Don't write to the object, leave it as it was until the user decides to make a change. Else you may end up fighting with whatever configuration management system they are using.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 2]
```diff
@@ -311,6 +311,12 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
 	// Fill all defaults explicitly
 	network.FillDefaults(&newOperConfig.Spec, prev, mtu)
 
+	klog.Infof("IPsec: bootstrap: ipsecConfig: %v", newOperConfig.Spec.DefaultNetwork.OVNKubernetesConfig.IPsecConfig)
+	klog.Infof("IPsec: bootstrap: prev: %v", prev.DefaultNetwork.OVNKubernetesConfig.IPsecConfig)
```
Comment (JoelSpeed) [inline pkg/controller/operconfig/operconfig_controller.go:315]: Can you explain the purpose of the bootstrap logic to me? I'm trying to understand what it's trying to achieve/what problem it solves

So far I've got to:
* Every time the operator starts, it creates a default `Network` object with all of the defaults populated
* It then applies this (using server side apply?) to the cluster?

How does this work for values that already exist? The `FillDefaults` doesn't modify anything that doesn't already exist?

Why do we need this up conversion? The API server should fill out the defaults for you if any have changed

**Comment by anuragthehatter** [comment]
Just want to update here . It seems the modes are not being supported runtime on this PR i.e. #2144 bot build. Steps I did:
1. Performed an upgrade from 4.14 (EW+NS) cluster to 2144 bot build
2. oc edit network.operator to change mode from Full to Disabled or External ( On a cluster EW and NS both configured)
3. There is no impact of current config and I see following in CNO logs

```
I0105 15:52:21.965258       1 operconfig_controller.go:314] IPsec: bootstrap: ipsecConfig: &{Disabled}
I0105 15:52:22.206413       1 ovn_kubernetes.go:547] IPsec:  Lopoking at ipsecConfig = &{Mode:Full}
I0105 15:52:22.206436       1 ovn_kubernetes.go:549] IPsec ipsecConfig = &{Mode:Full}
I0105 15:38:45.558295       1 operconfig_controller.go:314] IPsec: bootstrap: ipsecConfig: &{External}
I0105 15:38:45.931599       1 ovn_kubernetes.go:547] IPsec:  Lopoking at ipsecConfig = &{Mode:Full}
I0105 15:38:45.931624       1 ovn_kubernetes.go:549] IPsec ipsecConfig = &{Mode:Full}

```
Seems operconfig_controller is doing something but cluster falls back to Full.
Also `Lopoking` is a typo @jcaamano

**Comment by yuvalk** [comment]
@JoshSalomon 

take a look on b5bf935210c3cf8b3cd5d6ee69b2da8c0d5eea4c or 9efc5f1a3a47fb9a78d48f036a275522f9a4d15c
for api bump commit example

**Comment by anuragthehatter** [comment]
@yuvalk @JoshSalomon Changing mode from Full to External causes ds `ovn-ipsec-host` to go away. Is this expected? that ds shall only go away during Diasbled phase as far as I undrstand. Pls advise

**Comment by yuvalk** [comment]
> @yuvalk @JoshSalomon Changing mode from `Full` to `External` causes ds `ovn-ipsec-host` to go away. Is this expected? that ds shall only go away during `Disabled` phase as far as I undrstand. Pls advise

yes!
the DS is only for E-W, which is only enabled in full mode.
disabled = no ipsec at all
external = only n-s, ie - only the os extension
full = also e-w, ie - you'll also get the ds

**Comment by anuragthehatter** [comment]
> > @yuvalk @JoshSalomon Changing mode from `Full` to `External` causes ds `ovn-ipsec-host` to go away. Is this expected? that ds shall only go away during `Disabled` phase as far as I undrstand. Pls advise
> 
> yes! the DS is only for E-W, which is only enabled in full mode. disabled = no ipsec at all external = only n-s, ie - only the os extension full = also e-w, ie - you'll also get the ds

Thank you! and we will going to support thse object values from manifest as well (pre-install)?
```
defaultNetwork:
    type: OVNKubernetes
    ovnKubernetesConfig:
      ipsecConfig: {}                   <<<<<<<<It can be Full or external
```

**Comment by anuragthehatter** [comment]
Also whats the best way to check this on upgrade? I built a 4.14 cluster with EW+NS and upgraded it to bot build with this PR 2144 but ipsecConfig object values doesn't seem to be working on upgraded cluster.

**Pull request closed without merging**